### PR TITLE
Forward port 1010

### PR DESCRIPTION
**Pull request by d-maurer**
This ports #1010 to `master`.

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -15,6 +15,11 @@ https://github.com/zopefoundation/Zope/blob/4.x/CHANGES.rst
 
 - Update to newest compatible versions of dependencies.
 
+- Enhance cookie support. For details, see
+  `#1010 <https://github.com/zopefoundation/Zope/issues/1010>`_
+
+- Update dependencies to the latest releases that still support Python 2.
+
```
Comment (icemac) [inline CHANGES.rst:22]:
```suggestion
```

File: src/ZPublisher/tests/test_cookie.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+        def check(s):
+            ds = dump("name", s)
+            self.assertRegex(ds, r"[a-zA-Z0-9%_/]*$")
+            ls = load("name", ds)
+            self.assertEqual(ls, s)
+
+        check("abc")
+        check("a/b/c")
+        check(u"äöü")
```
Comment (icemac) [inline src/ZPublisher/tests/test_cookie.py:73]:
```suggestion
        check("äöü")
```

File: src/ZPublisher/tests/testHTTPResponse.py [hunk 3]
```diff
@@ -1499,7 +1486,8 @@ def encode(value, param):
                          1)
 
     def test_encode_words(self):
-        self.assertEqual(encode_words("ä"), "=?utf-8?b?w6Q=?=")
+        self.assertIn(encode_words(u"ä"),
```
Comment (icemac) [inline src/ZPublisher/tests/testHTTPResponse.py:1489]:
```suggestion
        self.assertIn(encode_words("ä"),
```

File: src/ZPublisher/tests/test_cookie.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        with self.assertRaises(ValueError):
+            convertCookieParameter("max_age", "abc")
+
+    def test_domain(self):
+        # already encoded
+        n, v = convertCookieParameter("domain", "xn--dmin-moa0i.example")
+        self.assertEqual(n, "Domain")
+        self.assertEqual(v, "xn--dmin-moa0i.example")
+        # perform encoding
+        _, v = convertCookieParameter("domain", u"dömäin.example")
```
Comment (icemac) [inline src/ZPublisher/tests/test_cookie.py:139]:
```suggestion
        _, v = convertCookieParameter("domain", "dömäin.example")
```

File: src/ZPublisher/tests/test_cookie.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        self.assertEqual(v, "xn--dmin-moa0i.example")
+        # check IDNA2003
+        _, v = convertCookieParameter("domain", u"Fußball.example")
```
Comment (icemac) [inline src/ZPublisher/tests/test_cookie.py:142]:
```suggestion
        _, v = convertCookieParameter("domain", "Fußball.example")
```

File: src/ZPublisher/tests/test_cookie.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        self.assertEqual(v, "fussball.example")
+        # check ``bytes`` handling
+        _, v = convertCookieParameter("domain",
+                                      u"Fußball.example".encode("utf-8"))
```
Comment (icemac) [inline src/ZPublisher/tests/test_cookie.py:146]:
```suggestion
                                      "Fußball.example".encode("utf-8"))
```

File: src/ZPublisher/tests/test_cookie.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        for t in "None Lax Strict".split():
+            n, v = convertCookieParameter("samesite", t.lower())
+            self.assertEqual(n, "SameSite")
+            self.assertEqual(v, t)
+        # bad value
+        with self.assertRaises(ValueError):
+            convertCookieParameter("samesite", "abc")
+
+    def test_comment(self):
+        n, v = convertCookieParameter("comment", u"comment")
```
Comment (icemac) [inline src/ZPublisher/tests/test_cookie.py:194]:
```suggestion
        n, v = convertCookieParameter("comment", "comment")
```

**Review by icemac — APPROVED** [review]
LGTM besides some `u` strings which are no longer necessary on Python 3-only.

**Comment by d-maurer** [comment]
The failing test is the `py39` build -- for this reason
```
Couldn't find index page for 'five.localsitemanager' (maybe misspelled?)
Getting distribution for 'five.localsitemanager==3.2.2'.
Couldn't find index page for 'five.localsitemanager' (maybe misspelled?)
While:
  Installing alltests.
  Getting distribution for 'five.localsitemanager==3.2.2'.
Error: Couldn't find a distribution for 'five.localsitemanager==3.2.2'.
```
I have no idea why it fails (the `py38` and `py310` builds succeed).

**Comment by jugmac00** [comment]
> The failing test is the `py39` build -- for this reason
> 
> ```
> Couldn't find index page for 'five.localsitemanager' (maybe misspelled?)
> Getting distribution for 'five.localsitemanager==3.2.2'.
> Couldn't find index page for 'five.localsitemanager' (maybe misspelled?)
> While:
>   Installing alltests.
>   Getting distribution for 'five.localsitemanager==3.2.2'.
> Error: Couldn't find a distribution for 'five.localsitemanager==3.2.2'.
> ```
> 
> I have no idea why it fails (the `py38` and `py310` builds succeed).

That is a transient error for sure - I manually restarted all builds.

**Review by dataflake — APPROVED** [review]

**Pull request merged**